src/bin/itmdump.rs

```diff
@@ -57,93 +60,95 @@ fn main() {
-    let (mut stdout, mut stderr) = (stdout.lock(), stderr.lock());
+    let stdout = io::stdout();
+    let mut stdout = stdout.lock();
     loop {
-        if let Err(e) = (|| {
-            try!(stream.read_exact(ref_slice_mut(&mut header)));
-            let port = header >> 3;
-
-            // Ignore all the packets that don't come from the stimulus port 0
-            if port != 0 {
-                return Ok(());
-            }
-
-            match header & 0b111 {
-                0b01 => {
-                    let mut payload = 0;
-                    try!(stream.read_exact(ref_slice_mut(&mut payload)));
-                    stdout.write_all(&[payload])
-                }
-                0b10 => {
-                    let mut payload = [0; 2];
-                    try!(stream.read_exact(&mut payload));
-                    stdout.write_all(&payload)
-                }
-                0b11 => {
-                    let mut payload = [0; 4];
-                    try!(stream.read_exact(&mut payload));
-                    stdout.write_all(&payload)
-                }
-                _ => {
-                    // Not a valid header, skip.
-                    Ok(())
+        let p = decoder.read_packet();
+        match p {
+            Ok(p) => {
+                match p.kind {
+                    packet::Kind::Instrumentation(ref ud) if ud.port == port => {
+                        stdout.write_all(&ud.payload)?;
+                    }
+                    _ => (),
                 }
             }
-        })() {
-            match e.kind() {
-                io::ErrorKind::UnexpectedEof => {
+            Err(e @ Error(ErrorKind::UnknownHeader(_), _)) => {
+                // We don't know this header type; warn and continue.
+                debug!("{}", e);
+            },
+            Err(Error(ErrorKind::EofBeforePacket, _)) => {
+                if follow {
                     thread::sleep(Duration::from_millis(100));
+                } else {
+                    // !follow and EOF. Exit.
+                    return Ok(())
                 }
-                _ => {
-                    writeln!(stderr, "error: {:?}", e.kind()).ok();
-                }
-            }
+            },
+
+            // FIXME: If in follow mode, we may try to read a packet
+            // but reach the end of the file in the middle. Currently
+            // we'd just return an error and hence exit. When we
+            // receive an error, we've already read and discarded some
+            // data, so we can't just go around this loop again.
+            //
+            // We could make a file following wrapper around `read`
+            // that blocks in a loop retrying the read and sleeping if
+            // there's no data.
+            Err(e) => return Err(e),
         }
-    }
+    } // end of read loop
+
+    // Unreachable.
+}
+
+fn open_read<'a>(matches: &ArgMatches) -> Result<impl io::Read + 'a> {
```

Comment:
Since you are creating a *boxed* trait object in both branches I think the return type should be `Box<Read + 'static>` instead of `impl Read`. Then the `impl_conservative_trait` feature gate can be dropped.

Reply:
Fixed. Yes, that was the fix I'd imagined.